src/compiler/crystal/semantic/semantic_visitor.cr

```diff
@@ -455,7 +455,12 @@ abstract class Crystal::SemanticVisitor < Crystal::Visitor
   def eval_macro(node, &)
     yield
   rescue ex : MacroRaiseException
-    node.raise ex.message, exception_type: MacroRaiseException
+    if ex.has_location?
+      # re-raise to retain the existing location
+      raise ex
+    else
+      node.raise ex.message, exception_type: MacroRaiseException
+    end
```

Comment:
Not sure why exactly, but when doing like:
```cr
macro macro_raise_on(arg)
  {% arg.raise "Oh noes" %}
end

macro_raise_on :this
```
The `:this` node doesn't have a location, so I made the logic conditional to only re-raise if it has a location, and only fall back onto a diff node if it doesn't. Otherwise this code example wouldn't have any line preview at all. Is a good fallback at least.

Reply:
Looks like the missing location for `SymbolLiteral` should be fixed though.